analyzers/src/SonarAnalyzer.CSharp/Rules/TypeMemberVisibility.cs

```diff
@@ -0,0 +1,94 @@
+    [Rule(DiagnosticId)]
+    public sealed class TypeMemberVisibility : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S3059";
+        private const string MessageFormat = "Types should not have members with visibility set higher than the type's visibility";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly SyntaxKind[] TypeKinds = { SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration, SyntaxKind.EnumDeclaration, SyntaxKindEx.RecordDeclaration };
+        private static readonly SyntaxKind[] MemberDeclarationKinds =
+        {
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.ConstructorDeclaration,
+            SyntaxKind.DelegateDeclaration,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.EventFieldDeclaration,
+            SyntaxKind.EnumDeclaration,
+            SyntaxKind.FieldDeclaration,
+            SyntaxKind.IndexerDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKindEx.RecordDeclaration,
+            SyntaxKind.StructDeclaration
+        };
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var typeDeclaration = (BaseTypeDeclarationSyntax)c.Node;
+                    var secondaryLocations = GetInvalidMemberLocations(typeDeclaration);
+                    if (secondaryLocations.Any())
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, typeDeclaration.Identifier.GetLocation(), secondaryLocations));
+                    }
+                },
+                TypeKinds);
+
+        private static List<Location> GetInvalidMemberLocations(BaseTypeDeclarationSyntax type)
+        {
+            var parentType = GetParentType(type);
+            if (parentType == null && type.Modifiers.AnyOfKind(SyntaxKind.InternalKeyword))
+            {
+                return type.DescendantNodes()
+                           .Where(node => node.IsAnyKind(MemberDeclarationKinds))
+                           .OfType<MemberDeclarationSyntax>()
+                           // We skip overridden methods since they need to keep the public visibility (if present)
+                           .Where(declaration => declaration.Modifiers().AnyOfKind(SyntaxKind.PublicKeyword) && !declaration.Modifiers().AnyOfKind(SyntaxKind.OverrideKeyword))
+                           .Select(declaration => declaration.Modifiers().Single(modifier => modifier.IsKind(SyntaxKind.PublicKeyword)).GetLocation())
+                           .ToList();
```

Comment:
Not a huge performance hit but maybe `ToList` conversion is not needed here.
It should work with return type `IEnumerable`.

Reply:
I've returned a list here to ensure that we have a single iteration. Otherwise both the call to `Any()` and `c.ReportDiagnosticWhenActive` will evaluate the Enumerable.